v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/handler.ts

```diff
@@ -0,0 +1,294 @@
+      const responses = await Promise.all(
+        jsonHttpRequest.map((singleReq: any) =>
+          this.#handleSingleRequest(singleReq),
+        ),
+      );
+
+      this.#sendResponse(res, responses);
+      return;
+    }
+
+    const rpcResp = await this.#handleSingleRequest(jsonHttpRequest);
+
+    this.#sendResponse(res, rpcResp);
+  };
+
+  public handleWs = async (ws: WebSocket): Promise<void> => {
+    const subscriptions: string[] = [];
+    let isClosed = false;
+
+    const listener = (payload: { subscription: string; result: any }) => {
+      // Don't attempt to send a message to the websocket if we already know it is closed,
+      // or the current websocket connection isn't interested in the particular subscription.
+      if (isClosed || !subscriptions.includes(payload.subscription)) {
+        return;
+      }
+
+      try {
+        ws.send(
+          JSON.stringify({
+            jsonrpc: "2.0",
+            method: "eth_subscription",
+            params: payload,
+          }),
+        );
+      } catch (error) {
+        _handleError(error);
+      }
+    };
+
+    // Handle eth_subscribe notifications.
+    this.#provider.addListener("notification", listener);
+
+    ws.on("message", async (msg: string) => {
+      let rpcReq: JsonRpcRequest | JsonRpcRequest[];
+      let rpcResp: JsonRpcResponse | JsonRpcResponse[];
+
+      try {
+        rpcReq = _readWsRequest(msg);
+
+        rpcResp = Array.isArray(rpcReq)
+          ? await Promise.all(
+              rpcReq.map((req) =>
+                this.#handleSingleWsRequest(req, subscriptions),
+              ),
+            )
+          : await this.#handleSingleWsRequest(rpcReq, subscriptions);
+      } catch (error) {
+        rpcResp = _handleError(error);
+      }
+
+      ws.send(JSON.stringify(rpcResp));
+    });
+
+    ws.on("close", () => {
+      // Remove eth_subscribe listener.
+      this.#provider.removeListener("notification", listener);
+
+      // Clear any active subscriptions for the closed websocket connection.
+      isClosed = true;
+      subscriptions.forEach(async (subscriptionId) => {
+        await this.#provider.request({
+          method: "eth_unsubscribe",
+          params: [subscriptionId],
+        });
+      });
+    });
+  };
+
+  #sendEmptyResponse(res: ServerResponse) {
+    res.writeHead(200);
+    res.end();
+  }
+
+  #setCorsHeaders(res: ServerResponse) {
+    res.setHeader("Access-Control-Allow-Origin", "*");
+    res.setHeader("Access-Control-Request-Method", "*");
+    res.setHeader("Access-Control-Allow-Methods", "OPTIONS, GET");
+    res.setHeader("Access-Control-Allow-Headers", "*");
+  }
+
+  #sendResponse(
+    res: ServerResponse,
+    rpcResp: JsonRpcResponse | JsonRpcResponse[],
+  ) {
+    res.statusCode = 200;
+    res.setHeader("Content-Type", "application/json");
+    res.end(JSON.stringify(rpcResp));
+  }
+
+  async #handleSingleRequest(req: JsonRpcRequest): Promise<JsonRpcResponse> {
```

Comment:
minor: I’d suggest moving the `handleRequest` function code into this function and renaming this one to `handleRequest`. Having both can be a bit confusing.